packages/next-upgrade/index.ts

```diff
@@ -0,0 +1,223 @@
+
+    const resLatest = await fetch(`https://registry.npmjs.org/next/latest`)
+    nextPackageJson['latest'] = await resLatest.json()
+  } catch (error) {
+    console.error('Failed to fetch versions from npm registry.')
+    return
+  }
+
+  let showRc = true
+  if (nextPackageJson['latest'].version && nextPackageJson['rc'].version) {
+    showRc =
+      compareVersions(
+        nextPackageJson['rc'].version,
+        nextPackageJson['latest'].version
+      ) === 1
+  }
+
+  const choices = [
+    {
+      title: 'Canary',
+      value: 'canary',
+      description: `Experimental version with latest features (${nextPackageJson['canary'].version})`,
+    },
+  ]
+  if (showRc) {
+    choices.push({
+      title: 'Latest Release Candidate',
+      value: 'rc',
+      description: `Pre-release version for final testing (${nextPackageJson['rc'].version})`,
+    })
+  }
+  choices.push({
+    title: 'Latest Stable',
+    value: 'latest',
+    description: `Production-ready release (${nextPackageJson['latest'].version})`,
+  })
+
+  const shortcutVersion = process.argv[2]?.replace('@', '')
+  if (['canary', 'rc', 'latest'].includes(shortcutVersion)) {
+    // Shortcut
+    targetVersionSpecifier = shortcutVersion as VersionSpecifier
+  } else {
+    const initialVersionSpecifierIdx = await processCurrentVersion()
+
+    const response: Response = await prompts({
+      type: 'select',
+      name: 'version',
+      message: 'What Next.js version do you want to upgrade to?',
+      choices: choices,
+      initial: initialVersionSpecifierIdx,
+    })
+
+    if (!response.version) {
+      return
+    }
+
+    targetVersionSpecifier = response.version
+  }
+
+  // TODO(LichuAcu): support Turborepo
+  const appPackageJsonPath = path.resolve(process.cwd(), 'package.json')
+  let appPackageJson = JSON.parse(fs.readFileSync(appPackageJsonPath, 'utf8'))
+
+  const targetNextPackageJson = nextPackageJson[targetVersionSpecifier]
+
+  if (targetNextPackageJson.version) {
+    if (compareVersions(targetNextPackageJson.version, '15.0.0-canary') >= 0) {
+      await suggestTurbopack(appPackageJson)
+    }
+  }
+
+  fs.writeFileSync(appPackageJsonPath, JSON.stringify(appPackageJson, null, 2))
+
+  // TODO(LichuAcu): keep the original styling of the file
+  console.log('\nUpdating depencies...\n')
+
+  let packageManager: PackageManager = getPackageManager(appPackageJson)
+  const dependencies = `react@${targetNextPackageJson.peerDependencies['react']} react-dom@${targetNextPackageJson.peerDependencies['react-dom']} next@${targetVersionSpecifier}`
+
+  let updateCommand
+  switch (packageManager) {
+    case 'pnpm':
+      updateCommand = `pnpm update ${dependencies}`
+      break
+    case 'npm':
+      updateCommand = `npm install ${dependencies}`
+      break
+    case 'yarn':
+      updateCommand = `yarn add ${dependencies}`
+      break
+    default:
+      updateCommand = `pnpm install next@${targetVersionSpecifier}`
+      break
```

Comment:
I think we should just ask the user (probably in `getPackageManager`) if we can't auto-determine the package manager. Running `pnpm` might break something.

We might be able to narrow the list by checking what binaries are in their `PATH`: https://github.com/npm/node-which/